<!DOCTYPE html>
<html class="writer-html5" lang="English" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Cell Neighborhoods Mapping and Functional Characterization &mdash; STIM 1.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
      <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css?v=61a4c737" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=1f8ec04b"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Inference of ligand-receptor interactions" href="LR_analysis.html" />
    <link rel="prev" title="Cross-modality integration of single-cell resolution Xenium in situ data and whole transcriptome Visium data" href="Integration.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            STIM
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Integration.html">Cross-Modality Integration of Single-Cell Resolution Xenium In Situ Data and Whole Transcriptome Visium Data</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Cell Neighborhoods Mapping and Functional Characterization</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#identify-cell-neighborhoods">Identify Cell Neighborhoods</a></li>
<li class="toctree-l3"><a class="reference internal" href="#proximity-enrichment-analysis">Proximity Enrichment Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#functional-analysis-of-neighborhoods">Functional Analysis of Neighborhoods</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="LR_analysis.html">Inference of ligand-receptor interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualizations.html">Core visualization functions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">STIM</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Cell Neighborhoods Mapping and Functional Characterization</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="cell-neighborhoods-mapping-and-functional-characterization">
<h1>Cell Neighborhoods Mapping and Functional Characterization<a class="headerlink" href="#cell-neighborhoods-mapping-and-functional-characterization" title="Link to this heading"></a></h1>
<p>High-resolution spatial transcriptomic data provides invaluable insights into cellular organization and interactions within their neighborhoods. These spatially organized neighborhoods play crucial roles in tissue functions and disease progression. This tutorial offers detailed instructions on how to perform cell neighborhood analysis using the package.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">from</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">import</span> <span class="n">rc_context</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">matplotlib.colors</span> <span class="k">as</span> <span class="nn">mcolors</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s1">&#39;WhRd&#39;</span><span class="p">,[</span><span class="s2">&quot;#e5e5e5&quot;</span><span class="p">,</span> <span class="s2">&quot;#fffacd&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;darkred&quot;</span><span class="p">],</span> <span class="n">N</span><span class="o">=</span><span class="mi">256</span><span class="p">)</span>  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pySTIM</span> <span class="k">as</span> <span class="nn">pst</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_vis</span> <span class="o">=</span> <span class="n">pst</span><span class="o">.</span><span class="n">load_Visium_data</span><span class="p">()</span>
<span class="n">adata_xe</span> <span class="o">=</span> <span class="n">pst</span><span class="o">.</span><span class="n">load_Xenium_data</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<section id="identify-cell-neighborhoods">
<h2>Identify Cell Neighborhoods<a class="headerlink" href="#identify-cell-neighborhoods" title="Link to this heading"></a></h2>
<p>We offers two methods to define cell neighborhoods: <strong>fixed radius</strong> or <strong>fixed K-nearest neighbors (KNN)</strong>. To specify the method, set the <code class="docutils literal notranslate"><span class="pre">mode</span></code> parameter to either '<strong>radius</strong>' or '<strong>knn</strong>'. You also need to specify the expected number of neighborhoods, which can be manually tuned by checking the cell type enrichment in neighborhoods. The calculated cell neighborhood results will be saved to <code class="docutils literal notranslate"><span class="pre">adata.obs[&quot;CN&quot;]</span></code>.</p>
<p>Using Radius-Based Method</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pst</span><span class="o">.</span><span class="n">calculate_cn</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;radius&#39;</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                <span class="n">group</span> <span class="o">=</span> <span class="s2">&quot;ident&quot;</span><span class="p">,</span> <span class="n">cluster</span> <span class="o">=</span> <span class="s2">&quot;celltype&quot;</span><span class="p">,</span>
                <span class="n">n_neighborhoods</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/a48cf75ba2ce3682b5fe6032a6ae65650f6c73b0e571b1ed5018e66eb4f5adff.png" src="../../_images/a48cf75ba2ce3682b5fe6032a6ae65650f6c73b0e571b1ed5018e66eb4f5adff.png" />
</div>
</div>
<p>Using KNN-based Method</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pst</span><span class="o">.</span><span class="n">calculate_cn</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;knn&#39;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> 
                <span class="n">group</span> <span class="o">=</span> <span class="s2">&quot;ident&quot;</span><span class="p">,</span> <span class="n">cluster</span> <span class="o">=</span> <span class="s2">&quot;celltype&quot;</span><span class="p">,</span>
                <span class="n">n_neighborhoods</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/ef8a410e79e707b66479a0b88c75bf728a19168767ee39e854637cd4e5d31d1f.png" src="../../_images/ef8a410e79e707b66479a0b88c75bf728a19168767ee39e854637cd4e5d31d1f.png" />
</div>
</div>
<p>Visualize the cell neighborhood organization in spatial.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;Set3&quot;</span><span class="p">,</span> <span class="n">lut</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="vertical-align: middle;"><strong>Set3</strong> </div><div class="cmap"><img alt="Set3 colormap" title="Set3" style="border: 1px solid #555;" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAgAAAABACAYAAABsv8+/AAAAE3RFWHRUaXRsZQBTZXQzIGNvbG9ybWFwH1kXwQAAABl0RVh0RGVzY3JpcHRpb24AU2V0MyBjb2xvcm1hcJTrh1kAAAAwdEVYdEF1dGhvcgBNYXRwbG90bGliIHYzLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZ59z+E8AAAAydEVYdFNvZnR3YXJlAE1hdHBsb3RsaWIgdjMuOC4wLCBodHRwczovL21hdHBsb3RsaWIub3JnsdVnaAAAAXpJREFUeJzt1rEtBWAYQFGegtIUJKIzlkb0bwR5Y7zEABqFRGMAI1DqRRQs8ceXuOdMcMt7ePf68nPwj9xefkwnLPP0eD6dsMzz59d0wjLbo910wjIPF8fTCcucfd9MJyzztn+fTljm9PpkOmGZq839dMJSm+kAAODvGQAACDIAABBkAAAgyAAAQJABAIAgAwAAQQYAAIIMAAAEGQAACDIAABBkAAAgyAAAQJABAIAgAwAAQQYAAIIMAAAEGQAACDIAABBkAAAgyAAAQJABAIAgAwAAQQYAAIIMAAAEGQAACDIAABBkAAAgyAAAQJABAIAgAwAAQQYAAIIMAAAEGQAACDIAABBkAAAgyAAAQJABAIAgAwAAQQYAAIIMAAAEGQAACDIAABBkAAAgyAAAQJABAIAgAwAAQQYAAIIMAAAEGQAACDIAABBkAAAgyAAAQJABAIAgAwAAQQYAAIIMAAAEGQAACDIAABBkAAAgyAAAQJABAIAgAwAAQb/00xDaSRv+wwAAAABJRU5ErkJggg=="></div><div style="vertical-align: middle; max-width: 514px; display: flex; justify-content: space-between;"><div style="float: left;"><div title="#8dd3c7ff" style="display: inline-block; width: 1em; height: 1em; margin: 0; vertical-align: middle; border: 1px solid #555; background-color: #8dd3c7ff;"></div> under</div><div style="margin: 0 auto; display: inline-block;">bad <div title="#00000000" style="display: inline-block; width: 1em; height: 1em; margin: 0; vertical-align: middle; border: 1px solid #555; background-color: #00000000;"></div></div><div style="float: right;">over <div title="#ffed6fff" style="display: inline-block; width: 1em; height: 1em; margin: 0; vertical-align: middle; border: 1px solid #555; background-color: #ffed6fff;"></div></div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cmap_CN</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;Set3&quot;</span><span class="p">,</span> <span class="n">lut</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">hex_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">rgb2hex</span><span class="p">(</span><span class="n">cmap_CN</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cmap_CN</span><span class="o">.</span><span class="n">N</span><span class="p">)]</span>
<span class="n">adata_xe</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;CN_colors&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hex_colors</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pst</span><span class="o">.</span><span class="n">plot_scatter</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">color_by</span><span class="o">=</span><span class="s2">&quot;CN&quot;</span><span class="p">,</span> <span class="n">ptsize</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">legend_col</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/a9e2fb00f23dff30863159c1d25a9100d5aacf279df1b3a910b16199be41689f.png" src="../../_images/a9e2fb00f23dff30863159c1d25a9100d5aacf279df1b3a910b16199be41689f.png" />
</div>
</div>
</section>
<section id="proximity-enrichment-analysis">
<h2>Proximity Enrichment Analysis<a class="headerlink" href="#proximity-enrichment-analysis" title="Link to this heading"></a></h2>
<p>Proximity enrichment analysis evaluates the enrichment frequency between cell type pairs using permutation-based tests. The significance of spatial relationships is determined by comparing observed frequencies to those from randomly permuted datasets. The output result is the enrichment z-scores. Note that this step may take a considerable amount of time (~10 minutes for 1000 permutations) when the cell number is large.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pst</span><span class="o">.</span><span class="n">calculate_enrichment</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="n">n_permutations</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> 
                              <span class="n">niche_radius</span> <span class="o">=</span> <span class="mi">15</span><span class="p">,</span> <span class="n">spatial_key</span> <span class="o">=</span> <span class="s2">&quot;spatial&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Total number of 85880 cells
Calculating observed interactions...
Permutation iterations: 0...
Permutation iterations: 100...
Permutation iterations: 200...
Permutation iterations: 300...
Permutation iterations: 400...
Permutation iterations: 500...
Permutation iterations: 600...
Permutation iterations: 700...
Permutation iterations: 800...
Permutation iterations: 900...
Finished!
</pre></div>
</div>
</div>
</div>
<p>Visualize the enrichment as heatmaps.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pst</span><span class="o">.</span><span class="n">plot_connectivity</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s2">&quot;Reds&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/fe3fd668218b2ee5d4b62594a53cb98bde4af8102bec0ed86ff1f7205a0ad7bb.png" src="../../_images/fe3fd668218b2ee5d4b62594a53cb98bde4af8102bec0ed86ff1f7205a0ad7bb.png" />
</div>
</div>
<p>Visualize the results as a graph network. In this network, each node represents a cell type, the node size indicates the cell type frequency, and edges connecting two cell types represent the enrichment scores. To enhance clarity, specify the cell types you want to highlight.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">color_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata_xe</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">celltype</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">,</span> <span class="n">adata_xe</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;celltype_colors&quot;</span><span class="p">]))</span>

<span class="n">pst</span><span class="o">.</span><span class="n">plot_network</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="n">color_dict</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s2">&quot;celltype&quot;</span><span class="p">,</span>  
                 <span class="n">highlight</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;PTS2&quot;</span><span class="p">,</span><span class="s2">&quot;PTS3&quot;</span><span class="p">,</span><span class="s2">&quot;Pod&quot;</span><span class="p">,</span><span class="s2">&quot;Glom-EC&quot;</span><span class="p">,</span><span class="s2">&quot;ICA&quot;</span><span class="p">,</span><span class="s2">&quot;ICB&quot;</span><span class="p">,</span><span class="s2">&quot;PC&quot;</span><span class="p">,</span><span class="s2">&quot;CNT&quot;</span><span class="p">],</span> <span class="n">seed</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/2c513a34a4660f5deb17035826bc8fb45f31ae7db80d5d8fdea358da509ee5d4.png" src="../../_images/2c513a34a4660f5deb17035826bc8fb45f31ae7db80d5d8fdea358da509ee5d4.png" />
</div>
</div>
</section>
<section id="functional-analysis-of-neighborhoods">
<h2>Functional Analysis of Neighborhoods<a class="headerlink" href="#functional-analysis-of-neighborhoods" title="Link to this heading"></a></h2>
<p>Next, we need to transfer the calculated CN labels to the Visium dataset for further analysis. Visium data provides whole transcriptome information, allowing for comprehensive functional analysis of the neighborhoods identified within the Xenium dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_vis</span> <span class="o">=</span> <span class="n">pst</span><span class="o">.</span><span class="n">add_metadata</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">adata_xe</span><span class="p">,</span> <span class="n">anchor_dict</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s2">&quot;CN&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Once the CN labels are transferred, compute the differential expression within the identified neighborhoods. This provide a more detailed molecular characterization compared to limited gene panel in Xenium dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;CN&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;wilcoxon&#39;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">de_df</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">group</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">marker_genes</span> <span class="o">=</span> <span class="n">de_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;group&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_vis</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;scaled&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">rc_context</span><span class="p">({</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="s1">&#39;figure.dpi&#39;</span><span class="p">:</span> <span class="mi">300</span><span class="p">}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="s1">&#39;CN&#39;</span><span class="p">,</span> <span class="n">dendrogram</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;scaled&#39;</span><span class="p">,</span> 
                               <span class="n">vmin</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/5da132232b9ebba4f8184a9fa75f957203914730664704c65d673808dbc6994f.png" src="../../_images/5da132232b9ebba4f8184a9fa75f957203914730664704c65d673808dbc6994f.png" />
</div>
</div>
<p>Additionally, perform gene set enrichment analysis to identify molecular signatures within specific neighborhoods.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">CN3_de</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">group</span> <span class="o">=</span> <span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="n">pval_cutoff</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">log2fc_min</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">names</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">enrichr_library</span> <span class="o">=</span> <span class="s1">&#39;GO_Biological_Process_2023&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pst</span><span class="o">.</span><span class="n">gesa</span><span class="p">(</span><span class="n">CN3_de</span><span class="p">,</span> <span class="n">enrichr_library</span><span class="p">,</span> <span class="n">show_number</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">bar_color</span> <span class="o">=</span> <span class="s1">&#39;#80b1d3&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/a818ed553d9b0df12c73f30b04e61a2ac608ef6e6a811d0cd43944bc866aefa7.png" src="../../_images/a818ed553d9b0df12c73f30b04e61a2ac608ef6e6a811d0cd43944bc866aefa7.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">enrichr_library</span> <span class="o">=</span> <span class="s1">&#39;MSigDB_Hallmark_2020&#39;</span>
<span class="n">pst</span><span class="o">.</span><span class="n">gesa</span><span class="p">(</span><span class="n">CN3_de</span><span class="p">,</span> <span class="n">enrichr_library</span><span class="p">,</span> <span class="n">show_number</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">bar_color</span> <span class="o">=</span> <span class="s1">&#39;#80b1d3&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/0cc80689592541358f1feb7e99734af4937c84163659df742025b3b391bceb92.png" src="../../_images/0cc80689592541358f1feb7e99734af4937c84163659df742025b3b391bceb92.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Integration.html" class="btn btn-neutral float-left" title="Cross-modality integration of single-cell resolution Xenium in situ data and whole transcriptome Visium data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="LR_analysis.html" class="btn btn-neutral float-right" title="Inference of ligand-receptor interactions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Humphreys Lab.</p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>