<!DOCTYPE html>
<html class="writer-html5" lang="English" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Cross-modality integration of single-cell resolution Xenium in situ data and whole transcriptome Visium data &mdash; STIM 1.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
      <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css?v=61a4c737" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=1f8ec04b"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Cell Neighborhoods Mapping and Functional Characterization" href="CN_analysis.html" />
    <link rel="prev" title="Tutorials" href="../tutorials.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            STIM
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Cross-modality integration of single-cell resolution Xenium in situ data and whole transcriptome Visium data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#aligning-spatial-transcriptomic-datasets">Aligning Spatial Transcriptomic Datasets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-histology-images-for-visualization">Adding histology images for visualization</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="CN_analysis.html">Cell Neighborhoods Mapping and Functional Characterization</a></li>
<li class="toctree-l2"><a class="reference internal" href="LR_analysis.html">Inference of ligand-receptor interactions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">STIM</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Cross-modality integration of single-cell resolution Xenium in situ data and whole transcriptome Visium data</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="cross-modality-integration-of-single-cell-resolution-xenium-in-situ-data-and-whole-transcriptome-visium-data">
<h1>Cross-modality integration of single-cell resolution Xenium in situ data and whole transcriptome Visium data<a class="headerlink" href="#cross-modality-integration-of-single-cell-resolution-xenium-in-situ-data-and-whole-transcriptome-visium-data" title="Link to this heading"></a></h1>
<p>This tutorial provides detailed instructions on how to integrate Xenium in situ data with Visium data performed on serial sections. It also includes functions for integrating high-resolution histology images, allowing visualization of both morphological details and molecular information.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">matplotlib.colors</span> <span class="k">as</span> <span class="nn">mcolors</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s1">&#39;WhRd&#39;</span><span class="p">,[</span><span class="s2">&quot;#e5e5e5&quot;</span><span class="p">,</span> <span class="s2">&quot;#fffacd&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;darkred&quot;</span><span class="p">],</span> <span class="n">N</span><span class="o">=</span><span class="mi">256</span><span class="p">)</span>  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pySTIM</span> <span class="k">as</span> <span class="nn">pst</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_vis</span> <span class="o">=</span> <span class="n">pst</span><span class="o">.</span><span class="n">load_Visium_data</span><span class="p">()</span>
<span class="n">adata_xe</span> <span class="o">=</span> <span class="n">pst</span><span class="o">.</span><span class="n">load_Xenium_data</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Load in Visium and Xenium data</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_xe</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;spatial&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata_xe</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;x_centroid&#39;</span><span class="p">,</span> <span class="s1">&#39;y_centroid&#39;</span><span class="p">]])</span> <span class="c1">## The original cell coords for Xenium data are stored in &#39;x_centroid&#39;, &#39;y_centroid&#39;</span>
<span class="n">adata_vis</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;spatial&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata_vis</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;Y_coords&#39;</span><span class="p">,</span> <span class="s1">&#39;X_coords&#39;</span><span class="p">]])</span> <span class="c1">## The original spot coords for Visium data are stored in &#39;X_coords&#39;, &#39;Y_coords&#39;</span>
</pre></div>
</div>
</div>
</div>
<p>We can visualize the marker gene expression of the sham kidneys from the two spatial transcriptomic platforms:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pst</span><span class="o">.</span><span class="n">plot_scatter</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">genes</span> <span class="o">=</span> <span class="s2">&quot;Slc7a13&quot;</span><span class="p">,</span> <span class="n">ptsize</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">legend_col</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cmap</span> <span class="o">=</span> <span class="n">cmap</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/8b5afeefc4632b0d88f9507a0a78dad50204806cf60eba309c2462f553fa6189.png" src="../../_images/8b5afeefc4632b0d88f9507a0a78dad50204806cf60eba309c2462f553fa6189.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pst</span><span class="o">.</span><span class="n">plot_scatter</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">genes</span> <span class="o">=</span> <span class="s2">&quot;Slc7a13&quot;</span><span class="p">,</span> <span class="n">ptsize</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">legend_col</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cmap</span> <span class="o">=</span> <span class="n">cmap</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/e9a19232886797e421c00121b7fdb48dddd50a4e0c698de5bcca2c08bf31fda5.png" src="../../_images/e9a19232886797e421c00121b7fdb48dddd50a4e0c698de5bcca2c08bf31fda5.png" />
</div>
</div>
<section id="aligning-spatial-transcriptomic-datasets">
<h2>Aligning Spatial Transcriptomic Datasets<a class="headerlink" href="#aligning-spatial-transcriptomic-datasets" title="Link to this heading"></a></h2>
<p>Several packages are available for aligning spatial transcriptomic datasets, including:<a class="reference external" href="https://paste-bio.readthedocs.io/en/latest/index.html">PASTE</a>, <a class="reference external" href="https://jef.works/STalign/">STalign</a>, and <a class="reference external" href="https://spatialdata.scverse.org/en/latest/index.html">SpatialData</a>. Since our datasets come from serial tissue sections, we aligned the two datasets through manual landmark selection using <a class="reference external" href="https://www.10xgenomics.com/support/software/xenium-explorer/latest/tutorials/xe-image-alignment">Xenium Explorer</a>. After registration, you can save the landmark points and transformation matrix for subsequent analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">H</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">0.09405798593520641</span><span class="p">,</span><span class="mf">3.41158653187994</span><span class="p">,</span><span class="o">-</span><span class="mf">959.2731416178372</span><span class="p">],</span>
             <span class="p">[</span><span class="mf">3.41158653187994</span><span class="p">,</span><span class="o">-</span><span class="mf">0.09405798593520641</span><span class="p">,</span><span class="o">-</span><span class="mf">1528.8859782583281</span><span class="p">],</span>
             <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">]])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aligned_points</span> <span class="o">=</span> <span class="n">pst</span><span class="o">.</span><span class="n">align_adata</span><span class="p">(</span><span class="n">H</span><span class="p">,</span> <span class="n">adata_xe</span><span class="p">,</span> <span class="n">adata_vis</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/a9d5edf6bf739ec95eecea52862a972fb71aa8c37c1d47b5bc1da33f2f6fdc88.png" src="../../_images/a9d5edf6bf739ec95eecea52862a972fb71aa8c37c1d47b5bc1da33f2f6fdc88.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_vis</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;spatial&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">aligned_points</span>
</pre></div>
</div>
</div>
</div>
<p>After alignment, we can obtain the alignment &quot;anchors&quot; which stores the mapping between Xenium cells and Visium spots. We can check the performance of cross-modality alignment by calculating the gene expression correspondance between Visium spots and Xenium pseudo-Visium spots.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">anchor_dict</span><span class="p">,</span> <span class="n">Xenium_agg_gene_df</span> <span class="o">=</span> <span class="n">pst</span><span class="o">.</span><span class="n">register_xenium_visium</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">adata_vis</span><span class="p">,</span> <span class="n">spot_diameter</span><span class="o">=</span><span class="mi">55</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Original Xenium cell #: 85880, mapped Xenium cell #: 23747
</pre></div>
</div>
<img alt="../../_images/7788ff13d45cd6f2f431fb27789995204da14229e187f05c0601158b6cf23e40.png" src="../../_images/7788ff13d45cd6f2f431fb27789995204da14229e187f05c0601158b6cf23e40.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Calculating pseudo-Visium data.....
Top 5 correlated spatial gene expression .....
         pearson_r
Acsm3     0.856231
Cyp7b1    0.829400
Slc12a1   0.789112
Bcat1     0.788783
Slc7a13   0.769708
Aqp1      0.767921
Slc14a2   0.748023
Ptger3    0.746897
Fgf1      0.746502
Epcam     0.745752
</pre></div>
</div>
</div>
</div>
<p>To visualize and compare the spatial gene expression patterns between the Visium and the pseudo-Visium data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s2">&quot;Acsm3&quot;</span>
<span class="n">pst</span><span class="o">.</span><span class="n">plot_compare</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">Xenium_agg_gene_df</span><span class="p">,</span> <span class="n">adata_vis</span><span class="p">,</span> <span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/c043895ee05e80cb69794b947c5127d84e0cb488e5357bfc679c43ebf09f6aef.png" src="../../_images/c043895ee05e80cb69794b947c5127d84e0cb488e5357bfc679c43ebf09f6aef.png" />
</div>
</div>
<p>Directly overlap Visium spot expression on Xenium cell expression.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s2">&quot;Slc12a1&quot;</span>
<span class="n">pst</span><span class="o">.</span><span class="n">plot_integrate</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">adata_xe</span><span class="p">,</span> <span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span> <span class="n">ptsize_vis</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">ptsize_xe</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color_bar</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/6a2c10cd3ada2314d04aa3d541e4f9106945a0c5adc18d9d0ec38df127940f4f.png" src="../../_images/6a2c10cd3ada2314d04aa3d541e4f9106945a0c5adc18d9d0ec38df127940f4f.png" />
</div>
</div>
</section>
<section id="adding-histology-images-for-visualization">
<h2>Adding histology images for visualization<a class="headerlink" href="#adding-histology-images-for-visualization" title="Link to this heading"></a></h2>
<p>To add pre-aligned morphology images, provide both the morphology image path and the DAPI image path, along with the affine transformation matrix/landmark points to the function <code class="docutils literal notranslate"><span class="pre">add_img()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_xe</span> <span class="o">=</span> <span class="n">add_img</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">img_path</span> <span class="o">=</span> <span class="s1">&#39;Xenium_shamR.tif&#39;</span><span class="p">,</span> <span class="n">dapi_path</span> <span class="o">=</span> <span class="s1">&#39;morphology.ome.tif&#39;</span><span class="p">,</span> <span class="n">transform</span> <span class="o">=</span> <span class="s1">&#39;Xenium_shamR_alignment_files/keypoints.csv&#39;</span><span class="p">,</span> <span class="n">scale_factor</span><span class="o">=</span><span class="mf">0.2125</span><span class="p">,</span> <span class="n">library_key</span><span class="o">=</span><span class="s2">&quot;spatial&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Visualize the cell type annotation or gene expression information on histology image data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pst</span><span class="o">.</span><span class="n">plot_scatter_img</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">[</span><span class="n">adata_xe</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">celltype</span> <span class="o">==</span> <span class="s2">&quot;Pod&quot;</span><span class="p">],</span> 
                 <span class="n">color_by</span> <span class="o">=</span> <span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="n">ptsize</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> 
                 <span class="n">image</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">image_key</span> <span class="o">=</span> <span class="s2">&quot;spatial&quot;</span><span class="p">,</span> 
                 <span class="n">seed</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">legend_col</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/51718c26d77cce15a0d2882d6197ccba76973a5fe652c7b0e45342dedd32cb5d.png" src="../../_images/51718c26d77cce15a0d2882d6197ccba76973a5fe652c7b0e45342dedd32cb5d.png" />
</div>
</div>
<p>To focus on a specific region of tissue morphology and molecular information, specify the coordinates of the desired region using the <code class="docutils literal notranslate"><span class="pre">x_lims</span></code> and <code class="docutils literal notranslate"><span class="pre">y_lims</span></code> parameters in the <code class="docutils literal notranslate"><span class="pre">plot_scatter_img()</span></code> function. In this example, we visualize the expression of <em>Nphs2</em>, a typical marker for podocytes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cmap_greens</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s2">&quot;Greens&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;#e5e5e5&quot;</span><span class="p">,</span> <span class="s2">&quot;#a1d99b&quot;</span><span class="p">,</span> <span class="s2">&quot;#31a354&quot;</span><span class="p">,</span> <span class="s2">&quot;#004a1b&quot;</span><span class="p">])</span>

<span class="n">pst</span><span class="o">.</span><span class="n">plot_scatter_img</span><span class="p">(</span><span class="n">adata_xe</span><span class="p">,</span> <span class="n">xlims</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1500</span><span class="p">,</span><span class="mi">2500</span><span class="p">],</span> <span class="n">ylims</span> <span class="o">=</span> <span class="p">[</span><span class="mi">3400</span><span class="p">,</span><span class="mi">4400</span><span class="p">],</span>
                 <span class="n">genes</span> <span class="o">=</span> <span class="s2">&quot;Nphs2&quot;</span><span class="p">,</span> <span class="n">ptsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">cmap</span> <span class="o">=</span> <span class="n">cmap_greens</span><span class="p">,</span>
                 <span class="n">image</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">image_key</span> <span class="o">=</span> <span class="s2">&quot;spatial&quot;</span><span class="p">,</span> 
                 <span class="n">seed</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">legend_col</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../../_images/c4c1d35450872ce532ca3280b209e35e0f35db7c1a9f0d528a08db68d4f881db.png" src="../../_images/c4c1d35450872ce532ca3280b209e35e0f35db7c1a9f0d528a08db68d4f881db.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../tutorials.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="CN_analysis.html" class="btn btn-neutral float-right" title="Cell Neighborhoods Mapping and Functional Characterization" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Humphreys Lab.</p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>