# STIM 

STIM provides a comprehensive computational pipeline for the integration and analysis of spatial transcriptomics data. This tool is designed to integrate tissue histology with both high-definition spatial transcriptomics (e.g., Xenium) and transcriptome-wide spatial transcriptomics (e.g., Visium). STIM enables: 
- Mapping gene expression across different spatial transcriptomics platforms
- Analyzing cell neighborhoods to identify spatial patterns and interactions
- Inferring ligand-receptor interactions

### Installation

Installation using pip:\
`pip install pySTIM` 

### Import
`import pySTIM as pst`

### Usage
<br>
***Cross-Modality Integration of Single-Cell Resolution Xenium In Situ Data and Whole Transcriptome Visium Data***: https://github.com/qiaoxy0/STIM/blob/main/source/docs/notebooks/Integration.ipynb
<br>

<br>
***Cell Neighborhoods Mapping and Functional Characterization***: https://github.com/qiaoxy0/STIM/blob/main/source/docs/notebooks/CN_analysis.ipynb
<br>

<br>
***Inference of ligand-receptor interactions***: https://github.com/qiaoxy0/STIM/blob/main/source/docs/notebooks/LR_analysis.ipynb
<br>

<br>
***Core Visualization functions***: https://github.com/qiaoxy0/STIM/blob/main/source/docs/notebooks/Visualizations.ipynb
<br>

Please look at our [documentation](https://stim-tutorial.readthedocs.io/) with example usage. 

